File: src/lib/rag.ts
export type RagSourceType = "chapter" | "figure" | "term" | "theme" | "map";

export type RagChunk = {
  id: string;
  sourceType: RagSourceType;
  slug: string;
  title: string;
  heading: string;
  text: string;
  tags: string[];
};

export type RetrievedChunk = RagChunk & { score: number };

export type RagAnswer = {
  answer: string;
  citations: Array<Pick<RagChunk, "id" | "sourceType" | "slug" | "title" | "heading">>;
  weakMatch: boolean;
};

const THEME_ALIASES: Record<string, string[]> = {
  焦慮: ["焦躁", "不安", "比較", "壓力", "有待", "逍遙遊"],
  比較: ["競爭", "輸贏", "高低", "有待", "逍遙遊"],
  升遷: ["工作", "職場", "成功", "功名", "有待", "無名", "無用之用"],
  成功: ["成就", "功名", "表現", "有待", "無名", "無用之用"],
  死亡: ["死亡", "死", "喪親", "失去", "死生", "大宗師", "至樂"],
  財富: ["金錢", "賺錢", "富有", "資產", "無用之用", "惠子", "惠施"],
  無用: ["沒用", "價值", "實用", "無用之用", "惠子", "逍遙遊"],
};

function queryTerms(query: string): string[] {
  const compact = query.replace(/\s+/g, "").trim();
  const aliases = Object.entries(THEME_ALIASES)
    .filter(([theme, words]) => compact.includes(theme) || words.some((word) => compact.includes(word)))
    .flatMap(([theme, words]) => [theme, ...words]);
  const chineseBigrams = [...compact].flatMap((_, index, chars) =>
    index < chars.length - 1 ? [chars.slice(index, index + 2).join("")] : [],
  );
  const words = query.split(/[\s，。！？、：；「」『』（）()]+/).filter((word) => word.length > 1);
  return [...new Set([compact, ...words, ...aliases, ...chineseBigrams])].filter((term) => term.length > 1);
}

function occurrences(text: string, term: string): number {
  return text.split(term).length - 1;
}

/** 中文關鍵詞與主題別名的輕量檢索，不連線至外部服務。 */
export function retrieve(query: string, chunks: RagChunk[], limit = 5): RetrievedChunk[] {
  const terms = queryTerms(query);
  if (terms.length === 0) return [];

  return chunks
    .map((chunk) => {
      const title = `${chunk.title} ${chunk.heading}`;
      const tags = chunk.tags.join(" ");
      let score = 0;
      for (const term of terms) {
        score += occurrences(chunk.text, term);
        if (title.includes(term)) score += 5;
        if (tags.includes(term)) score += 4;
      }
      return { ...chunk, score };
    })
    .filter((chunk) => chunk.score > 0)
    .sort((a, b) => b.score - a.score || a.title.localeCompare(b.title, "zh-Hant"))
    .slice(0, limit);
}

function excerpt(text: string): string {
  return text.replace(/\s+/g, " ").trim().slice(0, 220).replace(/[，、；：\s]+$/, "") + "。";
}

/**
 * 僅拼接已檢索文字，不自行補充《莊子》外的知識。
 * TODO: 未來可接 LLM，但提示詞必須強制只以 retrieved chunks 為上下文並保留引用。
 */
export function answerFromChunks(query: string, chunks: RagChunk[]): RagAnswer {
  const results = retrieve(query, chunks, 4);
  const weakMatch = results.length === 0 || results[0].score < 3;

  if (results.length === 0) {
    return {
      answer:
        "這是本庫內容的檢索結果，不是把下列話偽託為莊子原話。目前找不到足以直接回答此問題的材料；可改以「逍遙遊」、「有待」、「無待」、「無用之用」或「大宗師」查詢，隨內容擴充後再回來檢索。",
      citations: [],
      weakMatch: true,
    };
  }

  const summaries = results.map(
    (chunk) => `〈${chunk.title}〉「${chunk.heading}」指出：${excerpt(chunk.text)}`,
  );
  const caveat = weakMatch
    ? "目前命中內容較少，以下僅提供相關線索，並非完整定論。"
    : "以下整理只根據本庫已命中的段落，供閱讀與延伸思考。";

  return {
    answer: `這是本庫內容的檢索整理，不是宣稱「莊子直接對你說」。${caveat}\n\n${summaries.join("\n\n")}`,
    citations: results.map(({ id, sourceType, slug, title, heading }) => ({
      id,
      sourceType,
      slug,
      title,
      heading,
    })),
    weakMatch,
  };
}

File: streamlit_app/lib/rag.py
from __future__ import annotations

from typing import Any


THEME_ALIASES: dict[str, list[str]] = {
    "焦慮": ["焦躁", "不安", "比較", "壓力", "有待", "逍遙遊"],
    "比較": ["競爭", "輸贏", "高低", "有待", "逍遙遊"],
    "升遷": ["工作", "職場", "成功", "功名", "有待", "無名", "無用之用"],
    "成功": ["成就", "功名", "表現", "有待", "無名", "無用之用"],
    "死亡": ["死亡", "死", "喪親", "失去", "死生", "大宗師", "至樂"],
    "財富": ["金錢", "賺錢", "富有", "資產", "無用之用", "惠子", "惠施"],
    "無用": ["沒用", "價值", "實用", "無用之用", "惠子", "逍遙遊"],
}


def _terms(query: str) -> list[str]:
    compact = query.replace(" ", "").strip()
    aliases: list[str] = []
    for theme, words in THEME_ALIASES.items():
        if theme in compact or any(w in compact for w in words):
            aliases.extend([theme, *words])
    bigrams = [compact[i : i + 2] for i in range(max(0, len(compact) - 1))]
    words = [w for w in query.replace("，", " ").replace("。", " ").split() if len(w) > 1]
    out = []
    seen = set()
    for t in [compact, *words, *aliases, *bigrams]:
        if len(t) > 1 and t not in seen:
            seen.add(t)
            out.append(t)
    return out


def retrieve(query: str, chunks: list[dict[str, Any]], limit: int = 5) -> list[dict[str, Any]]:
    terms = _terms(query)
    if not terms:
        return []
    scored: list[dict[str, Any]] = []
    for chunk in chunks:
        title = f"{chunk.get('title', '')} {chunk.get('heading', '')}"
        tags = " ".join(chunk.get("tags") or [])
        text = chunk.get("text") or ""
        score = 0
        for term in terms:
            score += text.count(term)
            if term in title:
                score += 5
            if term in tags:
                score += 4
        if score > 0:
            scored.append({**chunk, "score": score})
    scored.sort(key=lambda c: (-c["score"], c.get("title", "")))
    return scored[:limit]


def answer_from_chunks(query: str, chunks: list[dict[str, Any]]) -> dict[str, Any]:
    results = retrieve(query, chunks, 4)
    weak = len(results) == 0 or results[0]["score"] < 3
    if not results:
        return {
            "answer": "目前找不到足以直接回答的本庫材料。可改問「逍遙遊」「無待」「無用之用」。",
            "citations": [],
            "weakMatch": True,
        }

    def excerpt(text: str) -> str:
        t = " ".join(text.split())
        return (t[:220].rstrip("，、；： ") + "。") if len(t) > 220 else t

    lines = [
        f"〈{c['title']}〉「{c['heading']}」：{excerpt(c['text'])}" for c in results
    ]
    caveat = "命中較弱，僅供線索。" if weak else "以下僅根據本庫已命中段落整理。"
    return {
        "answer": f"這是本庫檢索整理，不是偽託莊子原話。{caveat}\n\n" + "\n\n".join(lines),
        "citations": [
            {
                "id": c.get("id"),
                "sourceType": c.get("sourceType"),
                "slug": c.get("slug"),
                "title": c.get("title"),
                "heading": c.get("heading"),
            }
            for c in results
        ],
        "weakMatch": weak,
    }
